src/Sylius/Bundle/UserBundle/Context/CustomerContext.php

```diff
@@ -40,10 +51,34 @@ public function __construct(SecurityContextInterface $securityContext)
      */
     public function getCustomer()
     {
-        if ($this->securityContext->getToken() && $this->securityContext->isGranted('IS_AUTHENTICATED_REMEMBERED')
-            && $this->securityContext->getToken()->getUser()
-        ) {
-            return $this->securityContext->getToken()->getUser()->getCustomer();
+        if ($user = $this->getUser()) {
+            return $user->getCustomer();
+        }
+
+        return $this->session->get('customer', null);
+    }
+
+    /**
+     * Stores customer in session (if user is not authenticated) or on currently logged in user.
+     *
+     * @param CustomerInterface $customer
+     */
+    public function setCustomer(CustomerInterface $customer)
+    {
+        if ($user = $this->getUser()) {
+            $user->setCustomer($customer);
```

Comment:
The `User` object shouldn't be able to exist without `Customer` and the user's customer should never be changed (setter is only used during creation and should never be used after the user was created).